### ISSUE TITLE
Tensorflow softmax is not updating, docu doesn't help?

### ISSUE THREAD

**Issue by jandevries123** [tensorflow/tensorflow#23456]
"DDPG Algorithme not updating the softmax"

Dear reader,

please have a look at my ddpg code and what my problem, thanks in advance.

Code :: https://hastebin.com/beroramufu.py

What is working: The environment, S and S_ are working as they should i've checked several times if they aren't the same and there is a clear diffrence between the two. i changed the action space to discrete action spaces with 3 actions, for the actor network i changed the output to softmax, since i believe that worked out better for me to use... The Actions are working like they should and the network gives no error, i added a small change that it (should) save the model and load it in later if needed. the Batch i changed just a little bit, i believe i didn't changed it too much in this version, but i atleast changed the "memory-size" to 10 instead of 10000 since that would be better for testing and i will defenitly change it up after testing to a more reasonable amount, same for the "LR-a" and "LR-c" values, just to see if it acctually learns (will come to this later). 

What isn't working: As i said already the code works smoothly without errors, it can do actions and it gets the reward given for those actions, i use the E.G print(action[0]) to see the softmax for the 3 actions in my stepfunction (i choose actions there with the argmax method). works all smooth, but when i look at the values of the return of the softmax i can see they clearly do not change overal (not even a bit) over time, i did several runs with 2500 episodes without any changes. you can clearly see that in the picture below that indicated the reward given for the episode, you can clearly the line being straight and only moving due to a bit luck, but no real learning. the beginning of the line has a bit more volume but thats just because the first values in an average are higher, i strongly believe that you understand why it is like that.

Picture :: https://cdn.discordapp.com/attachments/455970083617898496/505451154168217610/unknown.png

So i determined that there is no learning going on, i know i should then try to print out the loss and other important values but i didn't seem to mange to print that since i got errors that are almost never seen in Tensorflow, with sess.run and tf.print, nothing really worked out, but i personally believe that that is not need, since the softmax for all the actions and the chart with the rewards + average indicate already enough that its just not learning.

So my conclusions i made is that the network is not updating properly, somewhere it is not passing the correct updates or something is a malfunction, i am not sure and i know for sure that it's probaly on my end with the code, i don't doubt your code, but more mine, thats why i would like to see your opinion about this and if you maybe know what it's causing by... seeing it.... i am sorry i can't send it running. i must tell you already that the implemented python multithreading and some people warned of this that that could affect the algorithme and that that the reason is that it's not working, but i am fairly condifent that that is NOT the issue, since the algorithme is the main thread and the environment is the thread that is even located in an other file, the multithreading is only created so that my environment can run whiles the algorithme extracts data and feeds actions to it whiles it's running. Not to forget the reset function returns a flat array of just values, some hot encoded some not. 

i wanna thank you in advance for reading, your time, your reply and your knownledge, if i missed anything, please notify me. 

Jan

**Comment by alextp** [comment]
This should be a stackoverflow question.

**Comment by jandevries123** [comment]
Please be so kind to delete this issue for me, thank you in advance.
